Assets/MixedRealityToolkit.Providers/WindowsMixedReality/WindowsMixedRealityHandRecorder.cs

```diff
@@ -0,0 +1,83 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using Microsoft.MixedReality.Toolkit.Utilities;
+using Microsoft.MixedReality.Toolkit.WindowsMixedReality.Input;
+using System;
+using System.Collections.Generic;
+using UnityEngine;
+
+#if WINDOWS_UWP
+using System.IO;
+using System.Threading.Tasks;
+using Windows.Storage;
+using Windows.Storage.Pickers;
+using Windows.Storage.Provider;
+#endif
+
+namespace Microsoft.MixedReality.Toolkit.Input
+{
+
+    /// <summary>
+    /// Record joint positions of a hand and log them for use in simulated hands
+    /// </summary>
+    public class WindowsMixedRealityHandRecorder : MonoBehaviour
+    {
+        private static readonly int jointCount = Enum.GetNames(typeof(TrackedHandJoint)).Length;
+
+        public TrackedHandJoint ReferenceJoint = TrackedHandJoint.IndexTip;
+
+        private Vector3 offset = Vector3.zero;
+        private Handedness recordingHand = Handedness.None;
+
+        public void RecordLeftHandStart()
+        {
+            RecordHandStart(Handedness.Left);
+        }
+
+        public void RecordRightHandStart()
+        {
+            RecordHandStart(Handedness.Right);
+        }
+
+        private void RecordHandStart(Handedness handedness)
+        {
+            HandJointUtils.TryGetJointPose<WindowsMixedRealityArticulatedHand>(ReferenceJoint, handedness, out MixedRealityPose joint);
+            offset = joint.Position;
+            recordingHand = handedness;
+        }
+
+        public void RecordHandStop()
+        {
+            MixedRealityPose[] jointPoses = new MixedRealityPose[jointCount];
+            for (int i = 0; i < jointCount; ++i)
+            {
+                HandJointUtils.TryGetJointPose<WindowsMixedRealityArticulatedHand>((TrackedHandJoint)i, recordingHand, out jointPoses[i]);
+            }
+
+            ArticulatedHandPose pose = new ArticulatedHandPose();
+            pose.ParseFromJointPoses(jointPoses, recordingHand, Quaternion.identity, offset);
+
+            recordingHand = Handedness.None;
+
+            StoreRecordedHandPose(pose.ToJson());
+        }
+
+        #if WINDOWS_UWP
+        private static void StoreRecordedHandPose(string data)
+        {
+            string path = Path.Combine(Application.persistentDataPath, "ArticulatedHandPose.json");
```

Comment:
What if I want to record several different hand poses? Will these get overwritten? Might make sense to timestamp these, for example "HandRecording-20190507_163011.json"

Reply:
Yes, timestamps should be simple. That would also avoid having to delete the file every time, because this method doesn't overwrite them ...

Initially i tried to make it really nice and open a file selection dialog with the UWP methods, but that didn;t work at all. I just needed a quick and dirty solution to record poses for the hand sim, and put the recorder in here for future improvement. It's not polished enough to be a user-facing tool at this point, on top of the fact that the list of supported poses is hardcoded atm.